common/darts/__init__.py

```diff
@@ -0,0 +1,14 @@
+__author__ = 'Todd Young'
+__email__ = 'youngmt1@ornl.gov'
+__version__ = '0.1.0'
+
+from .architecture import Architecture
+from .modules.conv.network import ConvNetwork
+from .modules.linear.network import LinearNetwork
+
```

Comment:
@jmohdyusof , how should we handle paths like these? Normally when I am setting up a package, I can use relative imports like this. Does this need to change for how we handle paths in Candle?

Reply:
I think this is unchanged, we do the same thing in candle (look at common/candle/__init__.py). This puts everything in the candle namespace, so this should be similar. We just have to point at ../../common from the benchmark (and we want to fix this step by doing a pip install so we get the path automatically). Does this make sense?

Reply:
`common/candle/__init__.py` doesn't have any examples where there are nested Python packages. Is that going to be an issue? For example, in `common/darts/__init__.py` I would like to 

```python
from .modules.conv.network import ConvNetwork
```
The modules subpackage has the following structure:
```
...
├  ...
  modules
├──   __init__.py
├──   classifier.py
├──   conv
│  ├──   __init__.py
│  ├──   cell.py
│  ├──   mixed_layer.py
│  └──   network.py
├──   linear
│  ├──   __init__.py
│  ├──   cell.py
│  ├──   mixed_layer.py
│  └──   network.py
└──   operations
   ├──   __init__.py
   ├──   conv.py
   └──   linear.py
```
I am not quite sure how to test this without having all the pieces in place and ready to run a benchmark.

Reply:
Looking at the way we did `P3B5`, we have the following:
```python
file_path = os.path.dirname(os.path.realpath(__file__))
lib_path = os.path.abspath(os.path.join(file_path, '..'))
sys.path.append(lib_path)
lib_path2 = os.path.abspath(os.path.join(file_path, '..', '..', 'common'))
sys.path.append(lib_path2)
```
Is this the way that I should expect things be brought into scope? If so, I suppose I could add `darts` to `lib_path2` to test how imports would work for other parts of CANDLE?

Something like this:
```
lib_path2 = os.path.abspath(os.path.join(file_path, '..', '..', 'common', 'darts'))
```
is that right?

Reply:
I don't think this is an issue, since these are all relative paths from common. At worst you might need to explicitly do the imports in the top-level __init__.py. If you run a test it should fail pretty quickly if the imports don't work?

Reply:
Re: the path. Yes, you could add the darts path as you show there. In future we might want to put candle_lib as a subdirectory to common, or not. But your solution + 'import darts' should bring all your functions into darts.XXX scope.

Reply:
To test the above I tried adding a simple `test.py` in our `P3B5` directory:

`test.py`
```python
import os
import sys

file_path = os.path.dirname(os.path.realpath(__file__))
lib_path = os.path.abspath(os.path.join(file_path, '..'))
sys.path.append(lib_path)
lib_path2 = os.path.abspath(os.path.join(file_path, '..', '..', 'common', 'darts'))
sys.path.append(lib_path2)

from darts.architecture import Architecture
from darts import ConvNetwork
from darts.genotypes import Genotype
```

And in `common/darts/__init__.py`:

`__init__.py`
```python
from __future__ import absolute_import

from architecture import Architecture
from .modules.conv.network import ConvNetwork
from .modules.linear.network import LinearNetwork
```
A couple interesting things here:
1. setting `lib_path2` in this way means that we are ignoring the contents in `darts/__init__.py`. This is bringing every Python module at the `common/darts` level into scope. 
2. This does not allow us to import anything from nested python packages within DARTS. In the above example, I cannot do something like `from darts import ConvNetwork`. The relative import `from .modules.conv.network` in `__init__.py` does not work this way.

Reply:
Sorry, I think I misundestood. 
1) In test.py you should just add /common to the path, then 'import darts'. That should execute the darts/__init__.py 
2) I think this might need to be done at the darts/__init__.py level? With candle we wanted to just have a single import which put all the user functions in scope. Does that work?

Reply:
I think I am going to have to either manually add to `sys.path` in each of the DARTS `__init__.py` files or move flatten out all of the directories to the level of `common/darts`.

Reply:
I added that [test.py](https://github.com/yngtodd/Benchmarks/blob/develop/Pilot3/P3B5/test.py) to show you what I am trying. 

After updating `sys.path` and importing darts, I cannot get a hold of `darts.architecture.Architecture`. It fails at  [line 14](https://github.com/yngtodd/Benchmarks/blob/develop/Pilot3/P3B5/test.py#L14)

Reply:
As written, test.py failed for me earlier: 
(tf-gpu) [jamal@cn4021 P3B5]$ python test.py
Traceback (most recent call last):
  File "test.py", line 12, in <module>
    import darts
  File "/home/jamal/Code/ECP/CANDLE/DARTS/Benchmarks/common/darts/__init__.py", line 8, in <module>
    from architecture import Architecture
ModuleNotFoundError: No module named 'architecture'

changing _init__.py fixed that: note that darts/__init__.py is not at the same relative level as candle/__init__.py

-from architecture import Architecture
+from .architecture import Architecture

(tf-gpu) [jamal@cn4021 P3B5]$ python test.py
Traceback (most recent call last):
  File "test.py", line 12, in <module>
    import darts
  File "/home/jamal/Code/ECP/CANDLE/DARTS/Benchmarks/common/darts/__init__.py", line 9, in <module>
    from .modules.conv.network import ConvNetwork
  File "/home/jamal/Code/ECP/CANDLE/DARTS/Benchmarks/common/darts/modules/conv/__init__.py", line 1, in <module>
    from .cell import Cell
  File "/home/jamal/Code/ECP/CANDLE/DARTS/Benchmarks/common/darts/modules/conv/cell.py", line 6, in <module>
    from darts.modules.operations.convolution import ConvBlock, FactorizedReduce
ModuleNotFoundError: No module named 'darts.modules.operations.convolution'

Is this what you see?

Reply:
Yeah, it is failing with the relative imports in `__init__.py`.

Reply:
If I change `test.py` to the following:
```python
import os
import sys


file_path = os.path.dirname(os.path.realpath(__file__))
lib_path = os.path.abspath(os.path.join(file_path, '..'))
sys.path.append(lib_path)
lib_path2 = os.path.abspath(os.path.join(file_path, '..', '..', 'common','darts'))
sys.path.append(lib_path2)


import darts
from darts.architecture import Architecture
from darts.modules.conv.network import Network

print(Architecture)
print(ConvNetwork)
print(darts.Genotype)
```
it then can get a hold of `Architecture`, but fails to import `Network`.

Reply:
I seem to get it mostly to work by fixing some renamed files? 

(tf-gpu) [jamal@cn4021 P3B5]$ git diff
diff --git a/common/darts/__init__.py b/common/darts/__init__.py
index 05c0e81..0a2b3f9 100644
--- a/common/darts/__init__.py
+++ b/common/darts/__init__.py
@@ -5,7 +5,7 @@ __email__ = 'youngmt1@ornl.gov'
 __version__ = '0.1.0'


-from architecture import Architecture
+from .architecture import Architecture
 from .modules.conv.network import ConvNetwork
 from .modules.linear.network import LinearNetwork

diff --git a/common/darts/modules/conv/__init__.py b/common/darts/modules/conv/__init__.py
index 0b2287c..e8b50cc 100644
--- a/common/darts/modules/conv/__init__.py
+++ b/common/darts/modules/conv/__init__.py
@@ -1,4 +1,4 @@
 from .cell import Cell
-from .network import Network
+from .network import ConvNetwork
 from .mixed_layer import MixedLayer

diff --git a/common/darts/modules/conv/cell.py b/common/darts/modules/conv/cell.py
index 203b4fa..f902ce5 100644
--- a/common/darts/modules/conv/cell.py
+++ b/common/darts/modules/conv/cell.py
@@ -3,7 +3,7 @@ import torch.nn as nn

 from darts.api import Model
 from darts.modules.conv.mixed_layer import MixedLayer
-from darts.modules.operations.convolution import ConvBlock, FactorizedReduce
+from darts.modules.operations.conv import ConvBlock, FactorizedReduce

This then gives me: trying to figure out the Genotype error. 

(tf-gpu) [jamal@cn4021 P3B5]$ python test.py
<class 'darts.architecture.Architecture'>
<class 'darts.modules.conv.network.ConvNetwork'>
Traceback (most recent call last):
  File "test.py", line 16, in <module>
    print(darts.Genotype)
AttributeError: module 'darts' has no attribute 'Genotype'

Reply:
I think Network was renamed to ConvNetwork? See my earlier comment, no need for multiple imports in test (which is how we would like it to be, the user just need to import darts)

Reply:
Works now:

(tf-gpu) [jamal@cn4021 P3B5]$ python test.py
<class 'darts.architecture.Architecture'>
<class 'darts.modules.conv.network.ConvNetwork'>
<class 'darts.genotypes.Genotype'>

just had to add 

from .genotypes import Genotype

to darts/__init__.py

Reply:
Man, I realized that I had a pip install version of my Darts library in my env. When removing that, I am getting even less to import. Would you mind sharing the `darts/__init__.py` and the `test.py` that you got to work?

Reply:
I still don't think this is the way we want it, I strongly prefer a single 'import darts'. For me this works: note I don't need sub-paths in the darts namespace, but the print will show those correctly. 

`import os
import sys


file_path = os.path.dirname(os.path.realpath(__file__))
lib_path = os.path.abspath(os.path.join(file_path, '..'))
sys.path.append(lib_path)
lib_path2 = os.path.abspath(os.path.join(file_path, '..', '..', 'common',))
sys.path.append(lib_path2)


import darts

print(darts.Architecture)
#print(darts.architecture.Architecture)
print(darts.ConvNetwork)
print(darts.LinearNetwork)
print(darts.Genotype)
`
with __init__.py

`from __future__ import absolute_import

__author__ = 'Todd Young'
__email__ = 'youngmt1@ornl.gov'
__version__ = '0.1.0'


from .architecture import Architecture
from .genotypes import Genotype
from .modules.conv.network import ConvNetwork
from .modules.linear.network import LinearNetwork


__all__ = [
    "Architecture",
    "ConvNetwork",
    "LinearNetwork",
]`

Reply:
Cool, I can keep to that convention.

Reply:
Great, I think if we were more familiar with it, there could probably be some better use of the __init__.py in subdirectories, so darts/__init__.py could just 'import modules' etc. but functionally it doesn't matter too much: we just want stuff in the darts namespace with a single include from the code.